### PR TITLE
Add 'QueryJob.results' method

### LINKED ISSUE #2083: [BigQuery] Feature Request - An easy way to get a QueryResults from a QueryJob
Problem:

There isn't an easy way to fetch the row results of a `client.run_async_query(job_id, query)` call.

I have to do some yucky stuff to get it to work in a sample I'm working on.

```python
client = bigquery.Client()
query_job = client.run_async_query(str(uuid.uuid4()), query)
query_job.use_legacy_sql = False
query_job.begin()

wait_for_job(query_job)

query_results = bigquery.query.QueryResults('', client)
query_results._properties['jobReference'] = {
    'jobId': query_job.name,
    'projectId': query_job.project
}
```

Possible Solutions:

- Modify constructor for QueryResults or add a class function to take an optional job object.
- Add a method to QueryJob `get_query_results()` that does this same yuckiness (modifying the internal `_properties`)

I'm sure there are other solutions, too, which is why I'm asking for input rather than going off and trying to fix right away.

Reference:

- QueryJob: [https://github.com/GoogleCloudPlatform/gcloud-python/blob/master/gcloud/bigquery/job.py#L933][1]
- QueryResults: [https://github.com/GoogleCloudPlatform/gcloud-python/blob/master/gcloud/bigquery/query.py#L42][2]
- Docs: https://raw.githubusercontent.com/GoogleCloudPlatform/gcloud-python/master/docs/bigquery-usage.rst Notice the async query does not show fetching row results.

[1]: https://github.com/GoogleCloudPlatform/gcloud-python/blob/6c49c54e94f79ef544ef75e6d7f2470f04656d3c/gcloud/bigquery/job.py#L933
[2]: https://github.com/GoogleCloudPlatform/gcloud-python/blob/6c49c54e94f79ef544ef75e6d7f2470f04656d3c/gcloud/bigquery/query.py#L42

### PR DESCRIPTION
**Pull request by tseaver**
Closes #2083.

/cc @jonparrott, @tswast

### CODE REVIEW

File: google/cloud/bigquery/_helpers.py [hunk 1]
```diff
@@ -172,3 +172,55 @@ def _validate(self, value):
         """
         if value not in self.ALLOWED:
             raise ValueError('Pass one of: %s' ', '.join(self.ALLOWED))
+
+
+class UDFResource(object):
```
Comment (tswast) [inline google/cloud/bigquery/_helpers.py]: Was this meant to be in this change? Seems fine to me to move the UDFResource class, but does this refactor belong in the PR to add QueryJob.results?

File: google/cloud/bigquery/query.py [hunk 2]
```diff
@@ -65,6 +65,26 @@ def __init__(self, query, client, udf_resources=()):
         self.udf_resources = udf_resources
         self._job = None
 
+    @classmethod
+    def from_query_job(cls, job):
```
Comment (tswast) [inline google/cloud/bigquery/query.py]: 👍 

`QueryResults.from_query_job` feels right to me.

**Comment by daspecster** [comment]
Should there be something in the bigquery-usage doc about this?

File: google/cloud/bigquery/_helpers.py (hunk 1 above)
Comment (tseaver) [inline google/cloud/bigquery/_helpers.py]: @tswast the move was part of reducing circular dependencies between `query.py` and `job.py`.

Comment (daspecster) [inline google/cloud/bigquery/_helpers.py]: I agree that it probably should have been in a different PR. I don't think it's worth going back for though.
If you don't squash merge then we could always cherry-pick that commit if needed. Just my $0.02

**Pull request merged**